libjob_queue/src/lsf_driver.c

```diff
@@ -330,6 +330,90 @@ static void lsf_driver_assert_submit_method( const lsf_driver_type * driver ) {
 
 
 
+/*
+  The resource request string contains spaces, and when passed
+  through the shell it must be protected with \"..\"; this applies
+  when submitting to a remote lsf server with ssh. However when
+  submitting to the local workstation using a bsub command the
+  command will be invoked with the util_spawn() command - and no
+  shell is involved. In this latter case we must avoid the \"...\"
+  quoting.
+*/
+static char * construct_quoted_resource_string(lsf_driver_type * driver) {
+  stringlist_type * select_list = stringlist_alloc_new();
+  if (stringlist_get_size(driver->exclude_hosts) > 0) {
+    for (int i = 0; i < stringlist_get_size(driver->exclude_hosts); i++) {
+      char * exclude_host = util_alloc_sprintf("hname!='%s'",
+                                               stringlist_iget(driver->exclude_hosts, i));
+      stringlist_append_owned_ref(select_list, exclude_host);
+    }
+  }
+
+  char * req = NULL;
+
+  if (stringlist_get_size(select_list) == 0) {
+    if (driver->resource_request)
+      req = util_alloc_string_copy(driver->resource_request);
+  } else {
+    // select_list is non-empty
+    if (driver->resource_request != NULL) {
+      char * resreq = util_alloc_string_copy(driver->resource_request);
+      char * excludes_string = stringlist_alloc_joined_string(select_list, " && ");
+
+      /*
+        A resource string can be "span[host=1] select[A && B] bla[xyz]".
+        The blacklisting feature is to have select[hname!=bad1 && hname!=bad2].
+      */
+
+      char* pos = strstr(resreq, "select["); // find select[...]
+      if (pos != NULL) {
+        // add select string to existing select[...]
+        char* endpos = strstr(pos, "]");
+        if (endpos != NULL)
+          *endpos = ' ';
+        else
+          util_abort("%s could not find termination of select statement: %s", __func__, resreq);
+
+        // We split string into (before) "bla[..] bla[..] select[xxx_"
+        // and (after) "... bla[..] bla[..]". (we replaced one ']' with ' ')
+        // Then we make final string:  before + &&excludes] + after
+        int before_size = endpos - resreq;
+        char before[before_size + 1];
```

Comment:
This could alternatively be:
```C
char * before = util_alloc_substring_copy( resreq, 0, before_size );
char * after = util_alloc_string_copy(&resreq[before_size]);
``` 
then the fiddling with 0 termination and `strncpy()`can be avoided.